Creating Superbuild

TheodoreDavis opened the pull request:
Started with the cpp dependency code.
TODO:
- Use ADD_EXTERNAL() for CMake_Public
- Tell hello_world it depends on CMake_Public
- Use FIND_PACKAGE() inside Hello_world DIR to locate CMake_Public

Commit:
Add optional code dependancy for cmake_public

ryanmrichard commented on hello_world/include/hello-world.hpp:
```diff
@@ -1,4 +1,10 @@
-std::string say_hello();
\ No newline at end of file
+#ifdef USING_CMAKEPUBLIC_
+#include <CMakePublic/cmake_public.hpp>
```
Remove the `#ifdef` and move the CMakePublic header to the source file. `#ifdef` like that is useful for optional dependencies; for simplicity let's just make this a required dependency. The header should be in the source file because no part of it is actually used in declaring the public API of CMakeSuperBuild.

ryanmrichard commented on hello_world/src/hello-world.cpp:
```diff
@@ -1,3 +1,24 @@
+
 std::string say_hello() { return std::string("Hello, CMake superbuild!"); }
+
+int main() {
```
This is a library so there's no main function. If you want, for a unit test you can make an executable that links to the library.

ryanmrichard commented on hello_world/src/hello-world.cpp:
```diff
@@ -1,3 +1,24 @@
 #include "hello-world.hpp"
 
+#ifdef USING_CMAKEPUBLIC_
```
Remove the `#ifdef` here too for the same reason.

Commit:
- Update .gitignore for VSCode.
- Make CMakePublic Required and move #include to source file.

Commit:
- Edit top level CMakeLists to build CMakePublic from zip of current head.
- Removed /${CMAKE INSTALL_PREFIX} from end of stage_dir
- Note: Switching stage_directory to CMAKE_INSTALL_PREFIX causes CMakePublic to build in the default location.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -1,18 +1,32 @@
 # This is where we will install artifacts during the build stage
-set(stage_directory "${CMAKE_BINARY_DIR}/stage/${CMAKE_INSTALL_PREFIX}")
+# TODO: change stage_dir to CMAKE_INSTALL_DIR
+set(stage_directory "${CMAKE_BINARY_DIR}/stage")
```
this was correct as it was

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -1,18 +1,32 @@
+include(ExternalProject) # For ExternalProject_Add command
+
+ExternalProject_Add(CMakePublic
+    URL "https://github.com/CMaizeExamples/CMakePublic/archive/refs/heads/master.zip"
```
I would recommend just using the git feature of ExternalProject_Add

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -1,18 +1,32 @@
+
+set(CMAKE_BUILD_TYPE_INIT "Release")
+
+project(CMaizeSuperbuild VERSION 1.0.0 LANGUAGES CXX)
```
This should be the second line in file, it actually initializes a bunch of variables.

TheodoreDavis commented:
Should I remove the staging?

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -1,18 +1,32 @@
-    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/hello_world"
-    CMAKE_ARGS -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
-	       -DCMAKE_INSTALL_PREFIX=${stage_directory}
-    BUILD_ALWAYS 1
+# ExternalProject_Add(
+#     "${PROJECT_NAME}_core"
+#     SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/hello_world"
+#     CMAKE_ARGS -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+# 	       -DCMAKE_INSTALL_PREFIX=${stage_directory}
+#     BUILD_ALWAYS 1
+# )
```
If you commented this out for testing purposes that's fine, but you're going to need this (and it's going to need to come after the `ExternalProject_Add` call for CMakePublic)

ryanmrichard commented on hello_world/include/hello-world.hpp:
```diff
@@ -2,9 +2,5 @@
 #include <iostream>
```
I missed it the first time, but `iostream` should be in your source file too. You're only using `string` in the header.

TheodoreDavis replied:
Since I'm no longer Printing shouldn't I just remove it?

ryanmrichard replied:
yeah. That's a better option.

ryanmrichard commented:
In a super build you have an outer build cycle which drives multiple inner build cycles. Each inner build cycle must be run to completion (i.e. you need to configure, build, and install it; you usually don't test dependencies and testing of the whole project occurs during the outer build's test step). The staging directory is where the inner build cycles are installed during the outer build cycle's build step. The install step of the outer build cycle copies the staging directory to `CMAKE_INSTALL_PREFIX`. You never want to build directly in the final location; things should only be moved to final location after they have been built successfully.

TheodoreDavis commented:
Gotcha, I was trying to treat hello_world as a normal CMake build!

ryanmrichard commented:
From the perspective of CMakeSuperBuild, you can't treat `hello_world` as a normal CMake build because the things it depends on (from CMakePublic) won't be available until after the ExternalProject command for CMakePublic has run, which happens during the build step of the CMakeSuperBuild repo.

Commit:
Remove iostream

Commit:
Top Level CMakeList
- Uncommented the external add for Hello_World.
- Added dependacy for CMakePublic to Hello_World.
- ReAdded CMAKE_INSTALL_PREFIX to stage_directory.

Commit:
Update hello-world CMakeList
- Change find package to look in folder rather than file.
- Change find package to look for config files
- Link CMakePublic lib
- delete cmake_public:: namespace in helloword.cpp
- Make include relative for cmake_public

Commit:
Re-Added return for EOF

ryanmrichard commented on hello_world/CMakeLists.txt:
```diff
@@ -1,46 +1,55 @@
+project(HelloWorld LANGUAGES CXX)
+
+# Can be generalized to CMAKE_PREFIX_PATH for multiple dependencies
+set(CMAKE_PREFIX_PATH "${CMAKE_INSTALL_PREFIX}/lib/cmake")
```
This should be passed in via the `ExternalProject_Add` command. Also you very rarely want to "set" a variable which starts with CMAKE. Those are the variables end users will use to control the build and if you set them you overwrite whatever value the end user set it to.

Commit:
Mov where CMAKE_PREFIX_PATH is defined

Commit:
Update CMakeLists
- Update comments
- Chance DCMAKE_PREFIX_PATH to use the actual variable
- Append our cmake config dir to CMAKE_PREFIX_PATH

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -2,15 +2,16 @@ cmake_minimum_required(VERSION 3.5.1)
 set(CMAKE_BUILD_TYPE_INIT "Release")
 
+# Add cmake config location to paths to search
+list(APPEND CMAKE_PREFIX_PATH ${stage_directory}/lib/cmake)
```
You should be able to leave off the `lib/cmake` part. CMake has a set of platform-specific subfolders that it looks in below the prefix thus insulating you from needing to change the path depending on the platform.

TheodoreDavis replied:
To make this work I'd need to change CMakePublic to place its config file in INSTALL_PREFIX/INSTALL_LIBDIR/PROJECT_NAME or INSTALL_PREFIX/INSTALL_LIBDIR/PROJECT_NAME/cmake instead of INSTALL_PREFIX/INSTALL_LIBDIR/cmake.
I went ahead and tested this on a local version of CMakePublic I had. I also think it's cleaner than just having all the config files dumped into one folder.

ryanmrichard replied:
the config files for CMakePublic should get placed in `INSTALL_PREFIX/cmakepublic/INSTALL_LIBDIR/cmake`. The libraries should go in `INSTALL_PREFIX/cmakepublic/INSTALL_LIBDIR` and the include files should go in `INSTALL_PREFIX/cmakepublic/INSTALL_INCDIR` (or whatever the variable for includes is...). The point is there should be a directory with the name `cmakepublic` that acts as a namespace so that the files don't just get mixed in with `INSTALL_PREFIX/lib` etc. (on Linux systems INSTALL_PREFIX/lib and the like are usually reserved for the OS's package management)

ryanmrichard replied:
If it's not doing this already then that's an error with the CMakePublic repo and needs fixed.

Commit:
Changes to HelloWorld CMakeLists
- Removed use of lib_name
- Added hello-world explicitly for install paths

Commit:
Update header path

Commit:
update prefix path

Commit:
First pass at unit testing
- Added test.cxx no dependency yet
- Added tests/CMakeList to install executable to stage/Debug
- Updated root CMakeList to add test as external project

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -38,3 +39,25 @@ ExternalProject_Add(
 INSTALL(DIRECTORY "${stage_directory}/" DESTINATION ${CMAKE_INSTALL_PREFIX})
+
+# Add Testing
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```
What's this logic for? I'm used to something like:
```.cmake
include(CTest) # I can't remember if this is supposed to go inside the `if` statement or not...
if(BUILD_TESTING)
    # Test set-up
endif()
```

ryanmrichard commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+
+#find_package(HelloWorld CONFIG REQUIRED)
+
+add_executable(${PROJECT_NAME} test.cxx)
```
For consistency, I believe we've been using `.cpp` as the extension for source files.

ryanmrichard commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+add_executable(${PROJECT_NAME} test.cxx)
+#target_link_libraries(${PROJECT_NAME} PUBLIC HelloWorld::HelloWorld)
+
+install(
```
You usually don't install tests; rather, you run them wherever they are built.

Commit:
Refactor tests
- Remove install step from test by telling it to echo instead
- Rename test.cxx to .cpp
Had to specifiy full file path in add_test to work this way. Not sure how to avoid it.

Commit:
Add dependencies to the test
- Fixed args I accidentally deleted from superbuild_core
- Removed the `-` from HelloWorld's install path conflicted with find_package
- Test needed find package for both depenencies.

Commit:
Same as last commit - forgot to add

Commit:
Fix indentation on 54? Uncomment DEPENDS for test

ryanmrichard commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+project(SuperTest LANGUAGES CXX)
+
+find_package(CMakePublic CONFIG REQUIRED)
+find_package(HelloWorld CONFIG REQUIRED)
```
If `HelloWorld` is set-up correctly (something this unit test is ultimately testing) you shouldn't have to look for `CMakePublic` again.

ryanmrichard commented on CMakeLists.txt:
```diff
@@ -1,20 +1,61 @@
+# Add Testing
+include(CTest)
+enable_testing()
+
```
I'd still protect this `ExternalProject_Add` with the `if(BUILD_TESTING)` so that the test subproject is only built when the user requests it. My prior comments were pertaining to what scenario was the `PROJECT_NAME` logic for?

TheodoreDavis replied:
Gotcha, I saw it on an [example](https://cliutils.gitlab.io/modern-cmake/chapters/testing.html). Though now that I go back and read it the reason they provide isn't relavent to this.

Commit:
replaced if statement to make building of tests optional

Commit:
Added comments

Commit:
need to print to ctesttestfile but it's getting overwritten

Commit:
Fixed `include(CTest)` overwriting CTestTestfile in root cmakelist

Commit:
Changed test to test for output rather than returning a `0`.

Commit:
working on removing find_package(public) for tests

Commit:
same as last

Commit:
Create package for helloworld
- Created a template `Config.cmake.in` for package.
- added CMakePublic depenency to package.
- Deleted file name of install(EXPORT).
- Added packaged config line to read template.
- Added install(FILES) for config file.

TheodoreDavis commented:
I Will add comments in on Monday

ryanmrichard commented on hello_world/cmake/Config.cmake.in:
```diff
@@ -0,0 +1,6 @@
+include("${CMAKE_CURRENT_LIST_DIR}/HelloWorld-targets.cmake")
+
+include(CMakeFindDependencyMacro)
+find_dependency(CMakePublic)
```
After finding it you still need to add the CMakePublic target to the imported HelloWorld target.

Commit:
- Change install targets to use relative paths
- Change helloworldconfig to be placed in build dir install of install dir
-

Commit:
Update some comments

ryanmrichard committed:
works on my computer

TheodoreDavis committed:
Merge pull request #1 from CMaizeExamples/tweaks

Tweaks

Commit:
Changed preview markdown to relative link to the actual documentation
